tests/performance/valgrind_instrumentation.h

```diff
@@ -0,0 +1,157 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2016-2022 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_valgrind_instrumentation_h
+#define dealii_valgrind_instrumentation_h
+
+#include <unistd.h>
+#include <valgrind/callgrind.h>
+
+#include <cstdio>
+#include <fstream>
+#include <mutex>
+
+
+/**
+ * A simple wrapper around callgrind to query current instruction counter
+ * values. In order to make use of the functions in this namespace the
+ * executable has to be started with valgrind:
+ * @code
+ *   valgrind --tool=callgrind -q --combine-dumps=yes \
+ *            --callgrind-out-file=callgrind.out ./executable
+ * @endcode
+ */
+namespace callgrind_wrapper
+{
+  namespace internal
+  {
+    /*
+     * Ensure that we start up with valgrind instrumentation in a stopped
+     * state.
+     */
+    struct OnStartup
+    {
+      OnStartup()
+      {
+        CALLGRIND_STOP_INSTRUMENTATION;
+        CALLGRIND_ZERO_STATS;
+        return;
+      };
+    } on_startup;
+  } // namespace internal
+
+
+  /**
+   * Reset the current instruction counter to zero and start callgrind
+   * instrumentation.
+   */
+  DEAL_II_ALWAYS_INLINE
+  inline void
+  start_instrumentation()
+  {
+    /*
+     * Verify that we are running under valgrind and that valgrind is
+     * called with the right parameters.
+     */
+    static bool is_initialized = []() {
+      // Is this executable running under valgrind supervision?
+      if (RUNNING_ON_VALGRIND != 1)
+        return false;
+
+      // Start and stop valgrind instrumentation
+      std::remove("callgrind.out");
+      CALLGRIND_STOP_INSTRUMENTATION;
```

Comment:
maybe stop first before we remove the file?

Reply:
Good point, fixed.